src/r3/atomic-swap/evm/src/SwapVault.sol

```diff
@@ -0,0 +1,185 @@
+ * limitations under the License.                                             *
+ ******************************************************************************/
+
+pragma solidity 0.8.20;
+
+import "openzeppelin/token/ERC20/IERC20.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+import "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin/token/ERC721/IERC721.sol";
+import "openzeppelin/token/ERC1155/IERC1155.sol";
+import "openzeppelin/token/ERC721/utils/ERC721Holder.sol";
+import "openzeppelin/token/ERC1155/utils/ERC1155Holder.sol";
+
+library Utils {
+    function safeSupportsInterface(address addr, bytes4 interfaceId) external view returns (bool supported) {
+        (bool success, bytes memory result) = addr.staticcall(
+                abi.encodeWithSelector(IERC165.supportsInterface.selector, interfaceId)
+            );
+        
+        supported = success && result.length > 0 && abi.decode(result, (bool));
+    }
+}
+
+contract SwapVault is ERC721Holder, ERC1155Holder {
+
+    error ZERO_ADDRESS();
+    error COMMIT_TO_SELF();
+    error ALREADY_COMMITTED();
+    error INVALID_AMOUNT();
+    error INVALID_STATUS();
+    error INVALID_CLAIMER();
+    error UNSUPPORTED_INTERFACE();
+
+    event Commit(string indexed swapId, bytes32 holdHash);
+    event Transfer(string indexed swapId, bytes32 holdHash);
+    event Revert(string indexed swapId, bytes32 holdHash);
+
+    using Utils for address;
+
+    struct Commitment {
+        string swapId;
+        uint256 status;
+        address owner;
+        address recipient;
+        uint256 amount;
+        uint256 tokenId;
+        address tokenAddress;
+        uint256 signaturesThreshold;                
+        bytes32[] signatures;
+    }
+
+    mapping(string => Commitment) _committmentState;
+
+    function commit(string calldata swapId, address recipient, uint256 signaturesThreshold) external {
+        _commit(swapId, recipient, signaturesThreshold);
+    }
+
+    function commitWithToken(string calldata swapId, address tokenAddress, uint256 amount, address recipient, uint256 signaturesThreshold) external {
+        Commitment storage commitment =  _commit(swapId, recipient, signaturesThreshold);
+
+        commitment.amount = amount;
+        commitment.tokenAddress = tokenAddress;
+
+        SafeERC20.safeTransferFrom(IERC20(commitment.tokenAddress), msg.sender, address(this), commitment.amount);
+
+        emit Commit(swapId, commitmentHash(swapId));
+    }
+
+    function commitWithToken(string calldata swapId, address tokenAddress, uint256 tokenId, uint256 amount, address recipient, uint256 signaturesThreshold) external {
+        Commitment storage commitment = _commit(swapId, recipient, signaturesThreshold);
+
+        commitment.amount = amount;
+        commitment.tokenId = tokenId;
+        commitment.tokenAddress = tokenAddress;
+
+        if(IERC165(tokenAddress).supportsInterface(0x80ac58cd)) { // ERC721
```

Comment:
@r3domfox when working with Solidity it's not like when working with other languages, so where instinctively one would say "use a uint8 here, you don't need uint256" that does not apply with Solidity. Same with code duplication or partial duplication. And that is because of how the EVM works and the execution cost (in terms of Gas) of EVM code.

Reply:
Is this quantifiable? i.e. estimate the gas costs before and after extracting the common functionality? Online sources I've found don't say much about the overheads of breaking out common functionality into private utility functions, and it's possible the compiler may inline them anyway in some circumstances - worth checking?

Reply:
@r3domfox I would consider it optimization at this point, or tentative optimization. Since the code is still evolving I would not spend time, at this stage, doing optimizations. Certainly once the logic is completed and tested, it is worth trying to optimizing code to reduce redundancy while maintaining gas costs same or lower. Just too early right now.